File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/algebra/solver/ImplyEdge.java
package edu.cornell.cs.apl.viaduct.algebra.solver;

import edu.cornell.cs.apl.viaduct.algebra.HeytingAlgebra;
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;
import org.jgrapht.graph.DefaultEdge;

/** Maps incoming values {@code input} to {@code c.imply(input)}. */
final class ImplyEdge<A extends HeytingAlgebra<A>> extends DefaultEdge implements DataFlowEdge<A> {
  private final A antecedent;

  ImplyEdge(A antecedent) {
    this.antecedent = antecedent;
  }

  @Override
  public A propagate(A input) {
    return antecedent.imply(input);
  }

  @Override
  public String toString() {
    return antecedent + " → _";
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/algebra/solver/ImplyEdge.kt
package edu.cornell.cs.apl.viaduct.algebra.solver

import edu.cornell.cs.apl.viaduct.algebra.HeytingAlgebra
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge
import org.jgrapht.graph.DefaultEdge

/** Maps incoming values `input` to `c.imply(input)`. */
internal class ImplyEdge<A : HeytingAlgebra<A>>(private val antecedent: A) : DefaultEdge(), DataFlowEdge<A> {
    override fun propagate(input: A): A {
        return antecedent.imply(input)
    }

    override fun toString(): String {
        return "$antecedent → _"
    }
}
